### ISSUE TITLE
uploading large file: execution expired

### ISSUE THREAD

**Issue by Aguasvivas22** [GoogleCloudPlatform/google-cloud-ruby#876]
script to upload file.
```
require 'google/cloud'

    # Uploading package to GCS
    begin
      gcloud = Google::Cloud.new "my-bucket"
      storage = gcloud.storage retries: 3, timeout: 300
      bucket = storage.bucket "my-bucket"
      bucket.create_file "large_file.zip", "large_file.zip"
      puts "Uploaded to GCS"
    rescue Exception => e
      puts 'Uploading to GCS failed'
      puts e.message
      puts "Backtrace:\n\t#{e.backtrace.join("\n\t")}"
    end
```

When uploading 1.8 GB file.
```
$ time ruby push.rb
Uploading to GCS failed
execution expired
Backtrace:
	/Library/Ruby/Gems/2.0.0/gems/google-cloud-storage-0.20.0/lib/google/cloud/storage/service.rb:305:in `rescue in execute'
	/Library/Ruby/Gems/2.0.0/gems/google-cloud-storage-0.20.0/lib/google/cloud/storage/service.rb:303:in `execute'
	/Library/Ruby/Gems/2.0.0/gems/google-cloud-storage-0.20.0/lib/google/cloud/storage/service.rb:181:in `insert_file'
	/Library/Ruby/Gems/2.0.0/gems/google-cloud-storage-0.20.0/lib/google/cloud/storage/bucket.rb:601:in `create_file'
	new_push.rb:11:in `<main>'

real	1m29.165s
user	0m9.945s
sys	0m1.489s
```

with and without the 300s timeout, it always errors out after about 1m 30s.

this only started recently and we were using gcloud 0.10.0.
I've since upgraded to google-cloud-storage, 0.20.0, but getting the same result.

I'm not sure if there's even a connection made or if it breaks, hangs, etc...

Is there anything I can add to get some more verbosity? Or even better, has anyone encountered this issue before and has a solution :)

Thanks in advance folks.

**Comment by blowmage** [comment]
@Aguasvivas22 Thanks for opening the issue. I've been looking into it. Hope to update you on this later today.

**Comment by blowmage** [comment]
@Aguasvivas22 Can you tell me which version of google-api-client you are using?

**Comment by Aguasvivas22** [comment]
Hey @blowmage, thanks for looking into this.

I'm using google-api-client (0.9.12)

**Comment by blowmage** [comment]
Hi @Aguasvivas22, I found the problem. `google-cloud-storage` uses Google API Client for Ruby to connect to the Storage service, and there are two main timeout settings: `timeout_sec` and `open_timeout_sec`. Unfortunately we are only configuring one of the settings, but I have opened #878 to fix this. As soon as the build passes I'll cut new releases of the gems that need this configuration change.

Thanks for opening this issue and letting us know about this!

**Comment by blowmage** [comment]
Here is some data I've collected for the behavior of Google API Client's `timeout_sec` and `open_timeout_sec` on Ftorage file uploads. The default for `open_timeout_sec` was set to 20, so the ~90 second execution time you saw was the original request and the 3 retries. Basically, `timeout_sec` is the setting that gets used for most API requests, but `open_timeout_sec` is the setting that gets used for uploads.

|Timeout|Open Timeout|Retries|Time To Error|
|-------|------------|-------|-------------|
|`nil`|`nil`|0|120.39083|
|`nil`|10|0|10.415425|
|10|`nil`|0|10.383179|
|10|10|0|10.397669|
|10|10|1|21.953786|
|10|10|2|32.720855|
|10|10|3|47.042215|
|60|10|0|10.3961|
|60|10|1|21.189269|
|60|10|2|34.711874|
|60|10|3|48.969367|
|10|60|0|60.491947|
|10|60|1|121.339585|
|10|60|2|184.25002|
|10|60|3|248.793393|
|60|60|0|60.385152|
|60|60|1|122.170743|
|60|60|2|182.841937|
|60|60|3|249.795168|
|20|20|0|20.394391|
|20|20|1|41.315982|
|20|20|2|65.178405|
|20|20|3|87.467426|
|120|20|0|20.430337|
|120|20|1|42.172156|
|120|20|2|63.579017|
|120|20|3|87.207945|
|20|120|0|120.432097|
|20|120|1|241.88666|
|20|120|2|364.956052|
|20|120|3|486.9879|
|120|120|0|120.46482|
|120|120|1|241.748111|
|120|120|2|363.232764|
|120|120|3|487.406526|

**Comment by Aguasvivas22** [comment]
Thanks for the details @blowmage.

Follow-up question:

does "open_timeout_sec" determine the max time an upload can take? Asking since our uploads usually take over a minute so wondering why it only started failing a couple days ago.

Thanks

**Comment by blowmage** [comment]
The version Google API Client for Ruby was recently changed [starting in v0.12](https://github.com/GoogleCloudPlatform/google-cloud-ruby/releases/tag/v0.12.0). It introduced some changes to the HTTP stack that are likely responsible. 

I'm just waiting on the CI build to complete so I can release.

**Comment by Aguasvivas22** [comment]
Cool, well thanks a lot for jumping on this so quickly!

**Comment by blowmage** [comment]
This is fixed in `google-cloud-storage` v0.20.1. Please update and let us know if it fixes this for you.

**Comment by Aguasvivas22** [comment]
upgraded and confirmed.

thanks!
